File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/models/kernel_pool.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

KernelPoolVarType = Literal["C", "N"]


@dataclass(frozen=True)
class StepKernelPoolGdpool:
    op: Literal["kernel-pool.gdpool"]
    name: str
    start: int
    room: int


@dataclass(frozen=True)
class StepKernelPoolGipool:
    op: Literal["kernel-pool.gipool"]
    name: str
    start: int
    room: int


@dataclass(frozen=True)
class StepKernelPoolGcpool:
    op: Literal["kernel-pool.gcpool"]
    name: str
    start: int
    room: int


@dataclass(frozen=True)
class StepKernelPoolGnpool:
    op: Literal["kernel-pool.gnpool"]
    template: str
    start: int
    room: int


@dataclass(frozen=True)
class StepKernelPoolDtpool:
    op: Literal["kernel-pool.dtpool"]
    name: str


@dataclass(frozen=True)
class StepKernelPoolPdpool:
    op: Literal["kernel-pool.pdpool"]
    name: str
    values: list[float]


@dataclass(frozen=True)
class StepKernelPoolPipool:
    op: Literal["kernel-pool.pipool"]
    name: str
    values: list[int]


@dataclass(frozen=True)
class StepKernelPoolPcpool:
    op: Literal["kernel-pool.pcpool"]
    name: str
    values: list[str]


@dataclass(frozen=True)
class StepKernelPoolSwpool:
    op: Literal["kernel-pool.swpool"]
    agent: str
    names: list[str]


@dataclass(frozen=True)
class StepKernelPoolCvpool:
    op: Literal["kernel-pool.cvpool"]
    agent: str


@dataclass(frozen=True)
class StepKernelPoolExpool:
    op: Literal["kernel-pool.expool"]
    name: str

File: packages/py-parity-checking/src/run-tspice/domains/kernel-pool.ts
import type {
  KernelPoolFoundNumbers,
  KernelPoolFoundStrings,
  StepKernelPoolCvpool,
  StepKernelPoolDtpool,
  StepKernelPoolExpool,
  StepKernelPoolGcpool,
  StepKernelPoolGdpool,
  StepKernelPoolGipool,
  StepKernelPoolGnpool,
  StepKernelPoolPcpool,
  StepKernelPoolPdpool,
  StepKernelPoolPipool,
  StepKernelPoolSwpool,
  StepOutput,
} from "../../case-types.js";
import type { RunTspiceContext } from "../context.js";

type KernelPoolStep =
  | StepKernelPoolGdpool
  | StepKernelPoolGipool
  | StepKernelPoolGcpool
  | StepKernelPoolGnpool
  | StepKernelPoolDtpool
  | StepKernelPoolPdpool
  | StepKernelPoolPipool
  | StepKernelPoolPcpool
  | StepKernelPoolSwpool
  | StepKernelPoolCvpool
  | StepKernelPoolExpool;

function normalizeFoundNumbers(out: { found: false } | { found: true; values: number[] }): KernelPoolFoundNumbers {
  if (!out.found) {
    return { found: false };
  }

  return { found: true, values: out.values };
}

function normalizeFoundStrings(out: { found: false } | { found: true; values: string[] }): KernelPoolFoundStrings {
  if (!out.found) {
    return { found: false };
  }

  return { found: true, values: out.values };
}

/** Execute one `kernel-pool.*` workflow step in tspice. */
export function runKernelPoolStep(
  context: RunTspiceContext,
  step: KernelPoolStep,
): StepOutput {
  switch (step.op) {
    case "kernel-pool.gdpool":
      return {
        op: step.op,
        value: normalizeFoundNumbers(context.spice.raw.gdpool(step.name, step.start, step.room)),
      };

    case "kernel-pool.gipool":
      return {
        op: step.op,
        value: normalizeFoundNumbers(context.spice.raw.gipool(step.name, step.start, step.room)),
      };

    case "kernel-pool.gcpool":
      return {
        op: step.op,
        value: normalizeFoundStrings(context.spice.raw.gcpool(step.name, step.start, step.room)),
      };

    case "kernel-pool.gnpool":
      return {
        op: step.op,
        value: normalizeFoundStrings(context.spice.raw.gnpool(step.template, step.start, step.room)),
      };

    case "kernel-pool.dtpool": {
      const out = context.spice.raw.dtpool(step.name);
      if (!out.found) {
        return { op: step.op, value: { found: false } };
      }

      return {
        op: step.op,
        value: {
          found: true,
          n: out.n,
          type: out.type,
        },
      };
    }

    case "kernel-pool.pdpool":
      context.spice.raw.pdpool(step.name, step.values);
      return { op: step.op, value: null };

    case "kernel-pool.pipool":
      context.spice.raw.pipool(step.name, step.values);
      return { op: step.op, value: null };

    case "kernel-pool.pcpool":
      context.spice.raw.pcpool(step.name, step.values);
      return { op: step.op, value: null };

    case "kernel-pool.swpool":
      context.spice.raw.swpool(step.agent, step.names);
      return { op: step.op, value: null };

    case "kernel-pool.cvpool":
      return { op: step.op, value: context.spice.raw.cvpool(step.agent) };

    case "kernel-pool.expool":
      return { op: step.op, value: context.spice.raw.expool(step.name) };

    default: {
      const exhaustive: never = step;
      throw new Error(`Unhandled kernel-pool step: ${JSON.stringify(exhaustive)}`);
    }
  }
}
